postgres/server/init.sls

```diff
@@ -191,4 +211,12 @@ postgresql-running:
       - file: postgresql-pg_hba
       - file: postgresql-pg_ident
 
+# Restart the service where reloading is not sufficient
+# Currently when changes are made to `postgresql.conf`
+{%- if postgres.postgresconf or db_port %}
+postgresql-service-restart:
+  service.running:
+    - name: {{ postgres.service }}
+{%- endif %}
+
```

Comment:
@myii I like the idea of using `onchanges` requisite much better. We could do something like this.

```yaml
{%- if postgres.postgresconf or db_port %}

postgresql-service-restart:
  module.run:
    - name: service.restart
    - m_name: {{ postgres.service }}
    - onchanges:
      - file: postgresql-conf

{%- endif %}
```

Reply:
@vutny Of course that can be done.  I have a couple of questions:

1. I'm curious: while I've used `module.run` for my own formulas, I've been under the impression that it should only be used as a last resort, when there isn't an available state -- for my understanding, would you mind explaining the advantage of using this instead?
1. To save having to keep the Jinja (`{%- if postgres.postgresconf or db_port %}` in sync with the one being used above, how about moving this into that block above, i.e.:

```sls
{%- if postgres.postgresconf or db_port %}

postgresql-conf:
  file.blockreplace:
    - name: {{ postgres.conf_dir }}/postgresql.conf
    - marker_start: "# Managed by SaltStack: listen_addresses: please do not edit"
    - marker_end: "# Managed by SaltStack: end of salt managed zone --"
    - content: |
        {%- if postgres.postgresconf %}
        {{ postgres.postgresconf|indent(8) }}
        {%- endif %}
        {%- if db_port %}
        port = {{ db_port }}
        {%- endif %}
    - show_changes: True
    - append_if_not_found: True
    {#- Detect empty values (none, '') in the config_backup #}
    - backup: {{ postgres.config_backup|default(false, true) }}
    - require:
      - file: postgresql-config-dir
      {%- if db_port %}
      - file: postgresql-conf-comment-port
      {%- endif %}

  module.run:
    - name: service.restart
    - m_name: {{ postgres.service }}
    - onchanges:
      - file: postgresql-conf

{%- endif %}
```

Further advantages of less lines and keeping logically related states together.

Reply:
@myii and @vutny what about the apache-formula's way

instead module.run:
```
apache-reload:
  module.wait:
    - name: service.reload
    - m_name: {{ apache.service }}
```

and then have the following in the file state:
```
    - watch_in:
      - module: apache-reload
```

Reply:
or just use listen_in:
https://docs.saltstack.com/en/latest/ref/states/requisites.html#listen-listen-in

Reply:
@aboe76 Unfortunately, can't use `listen`/`listen_in` because the service needs to be restarted _before_ the `postgres.manage` states run.

Reply:
@aboe76 @vutny `module.wait` + `watch`/`watch_in` looks like an improvement, thanks for that @aboe76.  However, the other issues raised above still remain.  At the very least, where should this be positioned?  In the same `{%- if postgres.postgresconf or db_port %}` block as shown above or at the bottom next to `service.running`?

Reply:
@myii I think it is OK to group together `postgresql-conf` and future `postgres-restart` states, just better name them separately, because that's a good practice.

In case of initial installation, when PG cluster is not running yet, the restart state would simply bring it up. Later `postgresql-running` state will just do reload when necessary, that's fast and safe. And it would bring the service back in case of any earlier failures. But this will prevent from unnecessary restarts then there would be no changes to `postgresql.conf` file.

With that said, your approach is good. Please update the PR and I'm sure we'll merge it in.
Thank you for your contribution!